plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java

```diff
@@ -838,43 +842,81 @@ public String getPluginDownloadUrl(Plugin plugin) {
      *                     be null
      * @return true if download is successful, false otherwise
      */
-    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE")
     public boolean downloadToFile(String urlString, Plugin plugin, File fileLocation) {
+        return downloadToFile(urlString, plugin, fileLocation, DEFAULT_MAX_RETRIES);
+    }
+
+    /**
+     * Downloads a plugin from a url to a file.
+     *
+     * @param urlString    String url to download the plugin from
+     * @param plugin       Plugin object representing plugin to be downloaded
+     * @param fileLocation contains the temp file if the plugin is downloaded so that dependencies can be parsed from
+     *                     the manifest, if the plugin will be downloaded to the plugin download location, this will
+     *                     be null
+     * @param maxRetries   Maximum number of times to retry the download before failing
+     * @return true if download is successful, false otherwise
+     */
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE")
+    public boolean downloadToFile(String urlString, Plugin plugin, File fileLocation, int maxRetries) {
         File pluginFile;
         if (fileLocation == null) {
-            pluginFile = new File(refDir + SEPARATOR + plugin.getArchiveFileName());
+            pluginFile = new File(refDir, plugin.getArchiveFileName());
             System.out.println("\nDownloading plugin " + plugin.getName() + " from url: " + urlString);
         } else {
             pluginFile = fileLocation;
         }
-        try (CloseableHttpClient httpclient = HttpClients.createSystem()) {
-            HttpClientContext context = HttpClientContext.create();
-            HttpHead httphead = new HttpHead(urlString);
-            try (CloseableHttpResponse response = httpclient.execute(httphead, context)) {
-                HttpHost target = context.getTargetHost();
-                List<URI> redirectLocations = context.getRedirectLocations();
-                // Expected to be an absolute URI
-                URI location = URIUtils.resolve(httphead.getURI(), target, redirectLocations);
-                FileUtils.copyURLToFile(location.toURL(), pluginFile);
-            } catch (URISyntaxException | IOException e) {
-                logVerbose(String.format("Unable to resolve plugin URL %s, or download plugin %s to file",
-                        urlString, plugin.getName()));
-                return false;
+
+        boolean success = true;
+        for (int i = 0; i < maxRetries; i++) {
+            if (pluginFile.exists()) {
+                pluginFile.delete();
+            }
+
+            try (CloseableHttpClient httpclient = HttpClients.custom().useSystemProperties()
+                    .addInterceptorLast((HttpRequestInterceptor) (request, context) -> {
+                        throw new IOException("Planned");
```

Comment:
what's this do?

Reply:
This is from the recommendations for using retry that I found: http://www.javabyexamples.com/retrying-requests-using-apache-httpclient-4/

Reply:
it looks like you get an option of whether you want to retry or not and this this code does always retry on failure which i guess is fine but not terribly clear.

how about:
```suggestion
                        throw new IOException("Retry on any failure");
```